<a href="https://colab.research.google.com/github/hawkeyedatatsai/Feature-Engineering/blob/main/6_Feature_Engineering_Exercise_(Core).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Import the data the drop the 'casual' and 'registered' columns. These are redundant with your target, 'count'.

In [2]:
df = pd.read_csv('https://docs.google.com/spreadsheets/d/e/2PACX-1vROUXPkYUkX-2W7JbJ0-oNKaXzpg4NtmU9IeWEY6yFKm32ZEJOpRh_soHD4BeIcuHjYik3SEoXmkgwj/pub?output=csv')
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   datetime    10886 non-null  object 
 1   season      10886 non-null  int64  
 2   holiday     10886 non-null  int64  
 3   workingday  10886 non-null  int64  
 4   weather     10886 non-null  int64  
 5   temp        10886 non-null  float64
 6   atemp       10886 non-null  float64
 7   humidity    10886 non-null  int64  
 8   windspeed   10886 non-null  float64
 9   casual      10886 non-null  int64  
 10  registered  10886 non-null  int64  
 11  count       10886 non-null  int64  
dtypes: float64(3), int64(8), object(1)
memory usage: 1020.7+ KB
None


datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 0:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 1:00:00       1        0           0        1  9.02  13.635   
2  2011-01-01 2:00:00       1        0           0        1  9.02  13.635   
3  2011-01-01 3:00:00       1        0           0        1  9.84  14.395   
4  2011-01-01 4:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40  
2        80        0.0       5          27     32  
3        75        0.0       3          10     13  
4        75        0.0       0           1      1

In [3]:
df = df.drop(['casual', 'registered'], axis=1)
df.head()

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 0:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 1:00:00       1        0           0        1  9.02  13.635   
2  2011-01-01 2:00:00       1        0           0        1  9.02  13.635   
3  2011-01-01 3:00:00       1        0           0        1  9.84  14.395   
4  2011-01-01 4:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  count  
0        81        0.0     16  
1        80        0.0     40  
2        80        0.0     32  
3        75        0.0     13  
4        75        0.0      1

In [4]:
#make a copy of this df for future comparison
df2 = df.copy()

#Transform the 'datetime' column into a datetime type and use it to create 3 new columns in the data frame containing the:


*   Name of the Month
*   Name of the Day of the Week
*   Hour of the Day

- Make sure all 3 new columns are 'object' datatype so they can be one-hot encoded later.

- Drop the 'datetime' and 'season' columns. These are now redundant.

In [5]:
df['datetime'] = pd.to_datetime(df['datetime'])
df['month (name)'] = df['datetime'].dt.month_name()
df['day of week (name)'] = df['datetime'].dt.day_name()
df['hour'] = df['datetime'].dt.hour.astype(str)
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   datetime            10886 non-null  datetime64[ns]
 1   season              10886 non-null  int64         
 2   holiday             10886 non-null  int64         
 3   workingday          10886 non-null  int64         
 4   weather             10886 non-null  int64         
 5   temp                10886 non-null  float64       
 6   atemp               10886 non-null  float64       
 7   humidity            10886 non-null  int64         
 8   windspeed           10886 non-null  float64       
 9   count               10886 non-null  int64         
 10  month (name)        10886 non-null  object        
 11  day of week (name)  10886 non-null  object        
 12  hour                10886 non-null  object        
dtypes: datetime64[ns](1), float64(3), int64(6), ob

datetime  season  holiday  workingday  weather  temp   atemp  \
0 2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1 2011-01-01 01:00:00       1        0           0        1  9.02  13.635   
2 2011-01-01 02:00:00       1        0           0        1  9.02  13.635   
3 2011-01-01 03:00:00       1        0           0        1  9.84  14.395   
4 2011-01-01 04:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  count month (name) day of week (name) hour  
0        81        0.0     16      January           Saturday    0  
1        80        0.0     40      January           Saturday    1  
2        80        0.0     32      January           Saturday    2  
3        75        0.0     13      January           Saturday    3  
4        75        0.0      1      January           Saturday    4

In [6]:
df = df.drop(['datetime', 'season'], axis=1)
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   holiday             10886 non-null  int64  
 1   workingday          10886 non-null  int64  
 2   weather             10886 non-null  int64  
 3   temp                10886 non-null  float64
 4   atemp               10886 non-null  float64
 5   humidity            10886 non-null  int64  
 6   windspeed           10886 non-null  float64
 7   count               10886 non-null  int64  
 8   month (name)        10886 non-null  object 
 9   day of week (name)  10886 non-null  object 
 10  hour                10886 non-null  object 
dtypes: float64(3), int64(5), object(3)
memory usage: 935.6+ KB
None


holiday  workingday  weather  temp   atemp  humidity  windspeed  count  \
0        0           0        1  9.84  14.395        81        0.0     16   
1        0           0        1  9.02  13.635        80        0.0     40   
2        0           0        1  9.02  13.635        80        0.0     32   
3        0           0        1  9.84  14.395        75        0.0     13   
4        0           0        1  9.84  14.395        75        0.0      1   

  month (name) day of week (name) hour  
0      January           Saturday    0  
1      January           Saturday    1  
2      January           Saturday    2  
3      January           Saturday    3  
4      January           Saturday    4

#The temperatures in the 'temp' and 'atemp' columns are in Celsius. Use `.apply()` and a Lambda function to convert them to Fahrenheit.

In [7]:
def celsius_to_fahrenheit(celsius):
    fahrenheit = (celsius * 1.8) + 32
    return fahrenheit

df['temp'] = df['temp'].apply(celsius_to_fahrenheit)
df['atemp'] = df['atemp'].apply(celsius_to_fahrenheit)

print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   holiday             10886 non-null  int64  
 1   workingday          10886 non-null  int64  
 2   weather             10886 non-null  int64  
 3   temp                10886 non-null  float64
 4   atemp               10886 non-null  float64
 5   humidity            10886 non-null  int64  
 6   windspeed           10886 non-null  float64
 7   count               10886 non-null  int64  
 8   month (name)        10886 non-null  object 
 9   day of week (name)  10886 non-null  object 
 10  hour                10886 non-null  object 
dtypes: float64(3), int64(5), object(3)
memory usage: 935.6+ KB
None


holiday  workingday  weather    temp   atemp  humidity  windspeed  count  \
0        0           0        1  49.712  57.911        81        0.0     16   
1        0           0        1  48.236  56.543        80        0.0     40   
2        0           0        1  48.236  56.543        80        0.0     32   
3        0           0        1  49.712  57.911        75        0.0     13   
4        0           0        1  49.712  57.911        75        0.0      1   

  month (name) day of week (name) hour  
0      January           Saturday    0  
1      January           Saturday    1  
2      January           Saturday    2  
3      January           Saturday    3  
4      January           Saturday    4

#Create a new column, 'temp_variance' which shows how much warmer or colder the current temperature ('temp') is than the average temperate for that day of the year ('atemp').  If the current temperature is warmer than average ('atemp'), the value in 'temp_variance' should be positive. Drop the 'atemp' column.

In [8]:
df['temp_variance'] = df['temp'] - df['atemp']
df['temp_category'] = df['temp_variance'].apply(lambda x: 'warm' if x > 0 else 'cold')
df = df.drop(['atemp'], axis=1)
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   holiday             10886 non-null  int64  
 1   workingday          10886 non-null  int64  
 2   weather             10886 non-null  int64  
 3   temp                10886 non-null  float64
 4   humidity            10886 non-null  int64  
 5   windspeed           10886 non-null  float64
 6   count               10886 non-null  int64  
 7   month (name)        10886 non-null  object 
 8   day of week (name)  10886 non-null  object 
 9   hour                10886 non-null  object 
 10  temp_variance       10886 non-null  float64
 11  temp_category       10886 non-null  object 
dtypes: float64(3), int64(5), object(4)
memory usage: 1020.7+ KB
None


holiday  workingday  weather    temp  humidity  windspeed  count  \
0        0           0        1  49.712        81        0.0     16   
1        0           0        1  48.236        80        0.0     40   
2        0           0        1  48.236        80        0.0     32   
3        0           0        1  49.712        75        0.0     13   
4        0           0        1  49.712        75        0.0      1   

  month (name) day of week (name) hour  temp_variance temp_category  
0      January           Saturday    0         -8.199          cold  
1      January           Saturday    1         -8.307          cold  
2      January           Saturday    2         -8.307          cold  
3      January           Saturday    3         -8.199          cold  
4      January           Saturday    4         -8.199          cold

Use a predictive model of your choice and try to predict the 'count' of hourly bike-share users with both the original features and the engineered feature set you created.

Remember to drop the 'casual' and 'registered' columns from both versions before modeling.

Did these feature engineering choices improve your ability to predict the 'count'?

#Machine Learning - Regression

In [9]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.pipeline import make_pipeline
from sklearn.compose import make_column_transformer, make_column_selector
from sklearn.dummy import DummyRegressor
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error, mean_absolute_percentage_error
from sklearn import set_config
set_config(display='diagram')

In [10]:
## Evaluation function
def evaluate_model(true,pred):
    mae = mean_absolute_error(true, pred)
    mape = mean_absolute_percentage_error(true, pred)
    mse = mean_squared_error(true, pred)
    rmse = np.sqrt(mse)
    r2 = r2_score(true, pred)
    print(f' MAE: {mae}, \n MAPE: {mape}, \n MSE: {mse}, \n RMSE: {rmse:}, \n R2: {r2}')

##Engineered df

In [11]:
#no missing value from above then drop duplicate rows
df.duplicated().sum()
df = df.drop_duplicates()
df.duplicated().sum()

0

In [12]:
print(df.describe())
print("\n")
print(df.describe(include="object"))
print("\n")
df.info()

            holiday    workingday       weather          temp      humidity  \
count  10885.000000  10885.000000  10885.000000  10885.000000  10885.000000   
mean       0.028571      0.680937      1.418466     68.417266     61.886174   
std        0.166606      0.466135      0.633855     14.024360     19.245894   
min        0.000000      0.000000      1.000000     33.476000      0.000000   
25%        0.000000      0.000000      1.000000     57.092000     47.000000   
50%        0.000000      1.000000      1.000000     68.900000     62.000000   
75%        0.000000      1.000000      2.000000     79.232000     77.000000   
max        1.000000      1.000000      4.000000    105.800000    100.000000   

          windspeed         count  temp_variance  
count  10885.000000  10885.000000   10885.000000  
mean      12.799928    191.590721      -6.163667  
std        8.164723    181.144505       2.820024  
min        0.000000      1.000000     -14.481000  
25%        7.001500     42.000000

In [13]:
# split X and y, you are predicting price
y = df['count']
X = df.drop(columns=['count'])

# split training and test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [14]:
#on numeric
scaler = StandardScaler()
num_selector = make_column_selector(dtype_include="number")
num_tuple = (scaler, num_selector)

#on nominal
ohe = OneHotEncoder(sparse_output=False, handle_unknown="ignore")
nom_pipeline = make_pipeline(ohe, scaler)
nom_selector = make_column_selector(dtype_include="object")
nom_tuple = (nom_pipeline, nom_selector)

# Instantiate the make column transformer
engineered_preprocessor = make_column_transformer(num_tuple, nom_tuple, remainder="drop", verbose_feature_names_out=False)
# Fit to X train
engineered_preprocessor.fit(X_train)

ColumnTransformer(transformers=[('standardscaler', StandardScaler(),
                                 <sklearn.compose._column_transformer.make_column_selector object at 0x7f76f1d72950>),
                                ('pipeline',
                                 Pipeline(steps=[('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False)),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 <sklearn.compose._column_transformer.make_column_selector object at 0x7f76f1d733d0>)],
                  verbose_feature_names_out=False)

In [15]:
bagreg = BaggingRegressor(random_state = 42)
#engineered
engineered_bagreg_pipe = make_pipeline(engineered_preprocessor, bagreg)
engineered_bagreg_pipe.fit(X_train, y_train)
engineered_bt_train_pred = engineered_bagreg_pipe.predict(X_train)
engineered_bt_test_pred = engineered_bagreg_pipe.predict(X_test)
#evaluation
print('\nEngineered Bagging Tree Train Evaluation')
evaluate_model(y_train, engineered_bt_train_pred)
print('\n')
print('\nEngineered Bagging Tree Test Evaluation')
evaluate_model(y_test, engineered_bt_test_pred)


Engineered Bagging Tree Train Evaluation
 MAE: 20.026568050961657, 
 MAPE: 0.24121572389492624, 
 MSE: 1097.145216000249, 
 RMSE: 33.12318245580049, 
 R2: 0.9664487551280185



Engineered Bagging Tree Test Evaluation
 MAE: 52.09191202197264, 
 MAPE: 0.5934198019388045, 
 MSE: 6182.122248638893, 
 RMSE: 78.62647295052025, 
 R2: 0.8134429261152041


##UN-Engineered df

In [16]:
print(df2.info())
df2.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   datetime    10886 non-null  object 
 1   season      10886 non-null  int64  
 2   holiday     10886 non-null  int64  
 3   workingday  10886 non-null  int64  
 4   weather     10886 non-null  int64  
 5   temp        10886 non-null  float64
 6   atemp       10886 non-null  float64
 7   humidity    10886 non-null  int64  
 8   windspeed   10886 non-null  float64
 9   count       10886 non-null  int64  
dtypes: float64(3), int64(6), object(1)
memory usage: 850.6+ KB
None


datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 0:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 1:00:00       1        0           0        1  9.02  13.635   
2  2011-01-01 2:00:00       1        0           0        1  9.02  13.635   
3  2011-01-01 3:00:00       1        0           0        1  9.84  14.395   
4  2011-01-01 4:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  count  
0        81        0.0     16  
1        80        0.0     40  
2        80        0.0     32  
3        75        0.0     13  
4        75        0.0      1

In [17]:
df.duplicated().sum()
df = df.drop_duplicates()
df.duplicated().sum()

0

In [18]:
# split X and y, you are predicting price
y = df2['count']
X = df2.drop(columns=['count'])

# split training and test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [19]:
#on numeric
scaler = StandardScaler()
num_selector = make_column_selector(dtype_include="number")
num_tuple = (scaler, num_selector)

#on nominal
ohe = OneHotEncoder(sparse_output=False, handle_unknown="ignore")
nom_pipeline = make_pipeline(ohe, scaler)
nom_selector = make_column_selector(dtype_include="object")
nom_tuple = (nom_pipeline, nom_selector)

# Instantiate the make column transformer
unengineered_preprocessor = make_column_transformer(num_tuple, nom_tuple, remainder="drop", verbose_feature_names_out=False)
# Fit to X train
unengineered_preprocessor.fit(X_train)

ColumnTransformer(transformers=[('standardscaler', StandardScaler(),
                                 <sklearn.compose._column_transformer.make_column_selector object at 0x7f76f1d97be0>),
                                ('pipeline',
                                 Pipeline(steps=[('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False)),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 <sklearn.compose._column_transformer.make_column_selector object at 0x7f76f1d96fe0>)],
                  verbose_feature_names_out=False)

In [20]:
bagreg = BaggingRegressor(random_state = 42)
#unengineered
unengineered_bagreg_pipe = make_pipeline(unengineered_preprocessor, bagreg)
unengineered_bagreg_pipe.fit(X_train, y_train)
unengineered_bt_train_pred = unengineered_bagreg_pipe.predict(X_train)
unengineered_bt_test_pred = unengineered_bagreg_pipe.predict(X_test)

#evaluation
print('\nUnengineered Bagging Tree Train Evaluation')
evaluate_model(y_train, unengineered_bt_train_pred)
print('\n')
print('\nUnengineered Bagging Tree Test Evaluation')
evaluate_model(y_test, unengineered_bt_test_pred)


Unengineered Bagging Tree Train Evaluation
 MAE: 39.21387800097991, 
 MAPE: 0.7598573685952053, 
 MSE: 4121.8213584027435, 
 RMSE: 64.20141243308237, 
 R2: 0.8743685427769943



Unengineered Bagging Tree Test Evaluation
 MAE: 103.35444526083762, 
 MAPE: 1.8226873131597947, 
 MSE: 25245.475047759, 
 RMSE: 158.88824704099105, 
 R2: 0.23065020417550797


#Comparison Results

| Model | R^2 Training Score | R^2 Test Score | RMSE Training Score | RMSE Test Score |
| ----------- | ----------- | ----------- | ----------- | ----------- |
| Bagged Tree-unengineered data | 0.874 | 0.231 | 64.201 | 158.888 |
| Bagged Tree-engineered data | 0.966 | 0.813 | 33.123 | 78.626  |


- Engineered data has significant better metrics across the board.

- The most significant one is on R2 test score.

- RMSE scores on Engineered data are half of the UNEngineered data.